aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -231,10 +231,16 @@ void FunctionalTensorWrapper::replace_(const Tensor& other) {
     value_ = at::_to_copy(value_, c10::TensorOptions().dtype(dtype()).layout(layout()));
     TORCH_INTERNAL_ASSERT(!value_.key_set().has(c10::DispatchKey::Functionalize));
   }
-  mutation_counter_++;
-  if (!at::GradMode::is_enabled() || InferenceMode::is_enabled()) {
-    // This mutation happened under no_grad or inference_mode
-    mark_mutation_during_no_grad_or_inference_mode();
+  // if a mutation happens to a view under a no_grad,
+  // we won't call replace_() on the other alias until the alias is later used, which
+  // might not be until after the no_grad region is exited.
+  // Therefore, replace_() is not unconditionally safe to check the current no_grad state.
```

Comment:
"However, if this is a lazy regeneration, then it is guaranteed that we have already done the mutation for the storage alias (when we originally performed the mutation), so no counter update may be needed."

Also, we can overreport this way, right? E.g., if I do a mutation on alias and I immediately replace_ the view and base, that will be +2 on the counters. Seems harmless for what you're using it for, but seems worth noting.

Reply:
> "However, if this is a lazy regeneration, then it is guaranteed that we have already done the mutation for the storage alias (when we originally performed the mutation), so no counter update may be needed."

yep this is clearer, will add

> f I do a mutation on alias and I immediately replace_ the view and base, that will be +2 on the counters

Hmm I'm not sure I follow. If I mutate an FunctionalTensorWrapper that is a base of some other FunctionalTensorWrapper, then the current setup looks like:

```
base = torch.ones(2)
view = base.view(-1)
view.mul_(2) # updates "view"

return base + base # updates "base" view lazy regenerate
```


<img width="829" alt="image" src="https://github.com/pytorch/pytorch/assets/16311747/68b2486e-e30a-4c5f-9622-66122e773c6d">

Where the following would run:

(1) `view.mul_(2)`: this calls `replace_()` exactly once, on the "view" tensor.

(2) `return base + base`: this calls `base->sync_()`, which runs the lazy regenerate logic to regenerate `base` off of the underlying "updated base tensor" in the `FunctionalStorageImpl`. In this case, there are zero views in the view chain between the underlying storage and `base`.

Reply:
Oh, I thought we immediately ran off and updated base sync (because the view has access to base), then I guess this does not happen.